app/safecast/tasks/device_story/update_metadata.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module Tasks
+  module DeviceStory
+    class UpdateMetadata
+      class << self
+        def call
+          new.call
+        end
+      end
+
+      def call
+        each_metadata(parse(fetch(devices_uri))) do |device_urn, metadata|
+          begin
+            find_or_update_device(device_urn, metadata)
+          rescue ActiveRecord::RecordInvalid
+            warn "Could not update device metadata for #{device_urn}."
+          end
+        end
+      end
+
+      private
+
+      def find_or_update_device(device_urn, metadata)
```

Comment:
I think this PR is a good idea -- start small, and build up functionality.

However, after seeing the problems with realtime, I think that this method for updating metadata might cause database performance problems that we could easily avoid up front.

If I understand it correctly, some of the metadata changes very often, and some of the metadata changes infrequently.

I suggest that when parsing the metadata, we create two separate objects, splitting out the metadata into these two groups. We should cache these objects somewhere.

Each time we fetch new data from ttserve, we compare the cached objects against the new objects and figure out exactly what we need to update in the DB. Postgres now has upsert syntax, so we can roll up all of the updates into a single SQL statement and avoid having to run a lot of UPDATEs.

https://stackoverflow.com/questions/1109061/insert-on-duplicate-update-in-postgresql

Maybe we can change ttserve output as well to make this easier?

Reply:
On second thought, I may have been too excessive for the relatively small number of queries here.